lib/charms/mongodb/v0/mongo.py

```diff
@@ -0,0 +1,284 @@
+            allow server time to start up.
+
+        Raises:
+            ConfigurationError, ConfigurationError, OperationFailure
+        """
+        try:
+            for attempt in Retrying(stop=stop_after_delay(60), wait=wait_fixed(3)):
+                with attempt:
+                    # The ping command is cheap and does not require auth.
+                    self.client.admin.command("ping")
+        except RetryError:
+            return False
+
+        return True
+
+    def create_user(self, config: MongoConfiguration):
+        """Create user.
+
+        Grant read and write privileges for specified database.
+        """
+        self.client.admin.command(
+            "createUser",
+            config.username,
+            pwd=config.password,
+            roles=self._get_roles(config),
+            mechanisms=["SCRAM-SHA-256"],
+        )
+
+    def update_user(self, config: MongoConfiguration):
+        """Update grants on database."""
+        self.client.admin.command(
+            "updateUser",
+            config.username,
+            roles=self._get_roles(config),
+        )
+
+    def set_user_password(self, username, password: str):
+        """Update the password."""
+        self.client.admin.command(
+            "updateUser",
+            username,
+            pwd=password,
+        )
+
+    def create_role(self, role_name: str, privileges: dict, roles: dict = []):
+        """Creates a new role.
+
+        Args:
+            role_name: name of the role to be added.
+            privileges: privileges to be associated with the role.
+            roles: List of roles from which this role inherits privileges.
+        """
+        try:
+            self.client.admin.command(
+                "createRole", role_name, privileges=[privileges], roles=roles
+            )
+        except OperationFailure as e:
+            if not e.code == 51002:  # Role already exists
+                logger.error("Cannot add role. error=%r", e)
+                raise e
+
+    @staticmethod
+    def _get_roles(config: MongoConfiguration) -> List[dict]:
+        all_roles = supported_roles(config)
+        return list(chain.from_iterable(all_roles[role] for role in config.roles))
+
+    def drop_user(self, username: str):
+        """Drop user."""
+        self.client.admin.command("dropUser", username)
+
+    def get_users(self) -> Set[str]:
+        """Add a new member to replica set config inside MongoDB."""
+        users_info = self.client.admin.command("usersInfo")
+        return set(
+            [
+                user_obj["user"]
+                for user_obj in users_info["users"]
+                if re.match(r"^relation-\d+$", user_obj["user"])
+            ]
+        )
+
+    def get_databases(self) -> Set[str]:
+        """Return list of all non-default databases."""
+        databases = self.client.list_database_names()
+        return set([db for db in databases if db not in SYSTEM_DBS])
```

Comment:
This is actually way more efficient in two ways:
  * Don't create a temp list (you can avoid this by directly providing the iterator to `set()`, but we're going one step further after that).
  * Calling `set(…)` instead of `{…}` is extremely inefficient (You can have a [read here](https://madebyme.today/blog/python-dict-vs-curly-brackets/) if you want to know more).
```suggestion
        return {db for db in databases if db not in SYSTEM_DBS}
```